### PR TITLE
enh: prevent 2 blocks from sharing the same name

### PR DESCRIPTION
**Pull request by fontanierh**
Only map/reduce and while/end should be allowed to have the same name.

### CODE REVIEW

File: core/src/app.rs [hunk 1]
```diff
@@ -211,16 +212,24 @@ impl App {
...
-                    block_type_names.insert(block_type_name);
+            // check that blocks don't share the same name, except for `map/reduce` and `while/end`
+            if let Some(block_types) = block_types_by_name.get_mut(name) {
+                if block_types.len() > 1
+                    || !((block.block_type() == BlockType::End
+                        && block_types.contains(&BlockType::While))
+                        || (block.block_type() == BlockType::Reduce
+                            && block_types.contains(&BlockType::Map)))
+                {
+                    Err(anyhow!(
```
Comment (spolu) [inline core/src/app.rs:223]: Woow that's hard to parse.

```
if block_types.len() > 0 && 
  !(block.block_type() == BlockType::End && block_types.contains(&BlockType::While)) &&
  !(block.block_type() == BlockType::Reduce && block_types.contains(&BlockType::Map))
```
Seems easier to parse. The `> 1` is quite confusing because it makes it unclear what happens for the simple case of a block having the same name as a previous one?

**Review by spolu — CHANGES_REQUESTED** [review]
Generally LGTM but the condition is really confusing (I could believe that it's correct but I'm not convinced even after spending quite some time looking at it)

File: core/src/app.rs (hunk 1 above)
Comment (spolu) [reply core/src/app.rs:223]: I still don't see in the current code path how == 0 is handled :grimacing:

Comment (fontanierh) [reply core/src/app.rs:223]: If it's 0, then `let Some(block_types) = block_types_by_name.get_mut(name)` cannot be true

Comment (fontanierh) [reply core/src/app.rs:225]: Added some comments let me know if that looks OK to you

**Review by spolu — APPROVED** [review]
LGTM
